packages/i18n/src/index.js

```diff
@@ -1,172 +1,7 @@
- *
- * @param {string} text     Text to translate.
- * @param {string} [domain] Domain to retrieve the translated text.
- *
- * @return {string} Translated text.
- */
-export function __( text, domain ) {
-	return dcnpgettext( domain, undefined, text );
-}
-
-/**
- * Retrieve translated string with gettext context.
- *
- * @see https://developer.wordpress.org/reference/functions/_x/
- *
- * @param {string} text     Text to translate.
- * @param {string} context  Context information for the translators.
- * @param {string} [domain] Domain to retrieve the translated text.
- *
- * @return {string} Translated context string without pipe.
- */
-export function _x( text, context, domain ) {
-	return dcnpgettext( domain, context, text );
-}
-
-/**
- * Translates and retrieves the singular or plural form based on the supplied
- * number.
- *
- * @see https://developer.wordpress.org/reference/functions/_n/
- *
- * @param {string} single   The text to be used if the number is singular.
- * @param {string} plural   The text to be used if the number is plural.
- * @param {number} number   The number to compare against to use either the
- *                          singular or plural form.
- * @param {string} [domain] Domain to retrieve the translated text.
- *
- * @return {string} The translated singular or plural form.
- */
-export function _n( single, plural, number, domain ) {
-	return dcnpgettext( domain, undefined, single, plural, number );
-}
-
-/**
- * Translates and retrieves the singular or plural form based on the supplied
- * number, with gettext context.
- *
- * @see https://developer.wordpress.org/reference/functions/_nx/
- *
- * @param {string} single   The text to be used if the number is singular.
- * @param {string} plural   The text to be used if the number is plural.
- * @param {number} number   The number to compare against to use either the
- *                          singular or plural form.
- * @param {string} context  Context information for the translators.
- * @param {string} [domain] Domain to retrieve the translated text.
- *
- * @return {string} The translated singular or plural form.
- */
-export function _nx( single, plural, number, context, domain ) {
-	return dcnpgettext( domain, context, single, plural, number );
-}
-
-/**
- * Returns a formatted string. If an error occurs in applying the format, the
- * original format string is returned.
- *
- * @param {string}    format The format of the string to generate.
- * @param {...string} args   Arguments to apply to the format.
- *
- * @see http://www.diveintojavascript.com/projects/javascript-sprintf
- *
- * @return {string} The formatted string.
- */
-export function sprintf( format, ...args ) {
-	try {
-		return sprintfjs.sprintf( format, ...args );
-	} catch ( error ) {
-		logErrorOnce( 'sprintf error: \n\n' + error.toString() );
-
-		return format;
-	}
-}
+export { sprintf } from './sprintf';
+export { I18n } from './i18n';
+export * from './default-i18n';
```

Comment:
We're exporting `i18n` (instance) and `I18n` (class), that differ by the capitalization of the leading `i`. It may be desirable to change one of the names.

Reply:
Should we be exporting the default instance at all? 

For prior art, compare to `@wordpress/data`, which is very similar in implementation of having a "default instance", and being able to create separate registries. There are some clear inconsistencies:

&nbsp;|`@wordpress/data`|`@wordpress/i18n`
---|---|---
Instantiate|`createRegistry`|`new I18n`
Default Instance Members|`select`, etc|`__`, etc
Default Instance|_(Not exported)_|`i18n`

Reply:
**`createI18n` vs `new I18n`:** The `createI18n()` function feels more like functional programming style, with a regular function that returns a value of certain type. `new I18n` is a classic OOP style. I certainly prefer the former, just keep in mind that in the end the `I18n` name will continue to be one of the package exports. It might not be the executable constructor, but rather the `I18n` type. It's the return type of `createI18n` and can be imported and used:
```js
import { I18n } from '@wordpress/i18n';
```

Similarly, `@wordpress/data` exposes the `DataRegistry` type:
```js
import { DataRegistry, createRegistry } from '@wordpress/data';

const registry: DataRegistry = createRegistry( ... );
```

**Should we be exporting the default instance at all?** I think the question to ask is whether there is any API that takes the whole registry as a parameter and uses it. And the answer is yes. The React bindings use the registry object all the time. `useRegistry` returns it. The `RegistryProvider` passes the registry object as the context value, and `RegistryConsumer` receives it.

The `defaultRegistry` value is the default value of the `RegistryProvider` context. The current implementation [imports it from a private relative path](https://github.com/WordPress/gutenberg/blob/master/packages/data/src/components/registry-provider/context.js#L9).

Imagine that we move the React bindings to a separate `@wordpress/react-data` package. Then the `defaultRegistry` must become part of `@wordpress/data` public API, because the React bindings need to use it.

Taking the above into consideration, I think that `@wordpress/data` should start exporting the `defaultRegistry`. And I like the idea to make it the default export from `@wordpress/data`.

Reply:
I disagree with exporting the defaultRegistry. I think ideally the defaultRegistry doesn't exist at all and should be instanciated like any other regular registry by the user. The problem though is that we need it for WordPress extensibility and we need it for BC reasons.

Ideally, for me, we should remove the default instances from the packages and add them as inline scripts in WordPress. Package consumers will always use it properly this way while we keep the default instances in WordPress contexts.

Reply:
The default instances are the ones causing this issue https://github.com/WordPress/gutenberg/issues/8981 so I don't think we should make this worse for package consumers.

Reply:
I agree that the default singleton instances shouldn't be there. That applies both to `data` and `i18n`, and also to packages like `core-data` or `block-directory`. These packages statically register a new store in the default registry, instead than exporting constructors and requiring the package consumer to wire up the instances.

But is there a viable and realistic plan to remove the singletons completely? And on what time frame? Or are we stuck with them forever because of backward compatibility?

If the default instances are going to stay, they should receive the love they deserve -- including the `defaultRegistry` export.

Reply:
Realistically speaking, they are here to stay in WordPress for BC and extensibility reasons but this doesn't mean they should stay in the packages IMO.

Reply:
So, the `@wordpress/data` package published in NPM will no longer create and export the default registry (i.e., the `select`, `dispatch` and `registerStore` functions), but when enqueuing the `wp-data` script in WordPress, the `wp.data` global will contain them? So there is no longer a 1:1 correspondence between the `@wordpress/foo` package exports and the `wp.foo` global?

Reply:
yes and this is already the case for a lot of packages.

Reply:
See inline scripts here https://github.com/WordPress/WordPress/blob/master/wp-includes/script-loader.php

Reply:
Is there an outcome to this discussion? How should I proceed?

Reply:
IMO: Let's not export the default instance. If we need to later, we can add it.